azure/services/privatedns/spec.go

```diff
@@ -0,0 +1,188 @@
+		}
+		if !isManaged(zone.Tags, z.ClusterName) {
+			log.V(1).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zone.Name)
+			log.V(1).Info("Tag the DNS manually from azure to manage it with capz."+
+				"Please see https://capz.sigs.k8s.io/topics/custom-dns.html#manage-dns-via-capz-tool", "private DNS", zone.Name)
+			return nil, nil
+		}
+	}
+
+	return privatedns.PrivateZone{
+		Location: to.StringPtr(azure.Global),
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: z.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Additional:  z.AdditionalTags,
+		})),
+	}, nil
+}
+
+// ResourceName returns the name of the virtual network link.
+func (l LinkSpec) ResourceName() string {
+	return l.Name
+}
+
+// OwnerResourceName returns the zone name of the virtual network link.
+func (l LinkSpec) OwnerResourceName() string {
+	return l.ZoneName
+}
+
+// ResourceGroupName returns the name of the resource group of the virtual network link.
+func (l LinkSpec) ResourceGroupName() string {
+	return l.ResourceGroup
+}
+
+// Parameters returns the parameters for the virtual network link.
+func (l LinkSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.LinkSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		link, ok := existing.(privatedns.VirtualNetworkLink)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.VirtualNetworkLink", existing)
+		}
+		if !isManaged(link.Tags, l.ClusterName) {
+			log.V(2).Info("Skipping vnet link reconciliation for unmanaged vnet link", "vnet link",
+				l.Name, "private dns zone", l.ZoneName)
+			return nil, nil
+		}
+	}
+
+	return privatedns.VirtualNetworkLink{
+		VirtualNetworkLinkProperties: &privatedns.VirtualNetworkLinkProperties{
+			VirtualNetwork: &privatedns.SubResource{
+				ID: to.StringPtr(azure.VNetID(l.SubscriptionID, l.VNetResourceGroup, l.VNetName)),
+			},
+			RegistrationEnabled: to.BoolPtr(false),
+		},
+		Location: to.StringPtr(azure.Global),
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: l.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Additional:  l.AdditionalTags,
+		})),
+	}, nil
+}
+
+// ResourceName returns the name of the virtual network link.
+func (r RecordSpec) ResourceName() string {
+	return r.Record.Hostname
+}
+
+// OwnerResourceName returns the zone name of the virtual network link.
+func (r RecordSpec) OwnerResourceName() string {
+	return r.ZoneName
+}
+
+// ResourceGroupName returns the name of the resource group of the virtual network link.
+func (r RecordSpec) ResourceGroupName() string {
+	return r.ResourceGroup
+}
+
+// Parameters returns the parameters for the virtual network link.
+func (r RecordSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	set := privatedns.RecordSet{
```

Comment:
Do we want to check if `existing != nil` and make sure it's the right type, and return `nil, nil` to skip the create/update?

Reply:
yeah, makes sense.